### PR TITLE
search: add tarantool-ee version searching

### PR DESCRIPTION
**Pull request by 0x501D**
This patch adds `tt search tarantool-ee`.
Also added functions for working with versions,
including sorting.

### CODE REVIEW

**Comment by 0x501D** [comment]
Will split into two commits..

File: cli/config/config.go [hunk 1]
```diff
@@ -29,6 +31,12 @@ type ModulesOpts struct {
 	Directory string
 }
 
+// EEOpts is uses to store tarantool-ee options.
```
Comment (better0fdead) [inline cli/config/config.go:34]:
```suggestion
// EEOpts is used to store tarantool-ee options.
```

File: cli/install_ee/install_ee.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+		}
+		versions = append(versions, version)
+	}
+
+	version.SortVersions(versions)
+
+	return versions, nil
+}
+
+func FetchVersions(cliOpts *config.CliOpts) ([]version.Version, error) {
```
Comment (better0fdead) [inline cli/install_ee/install_ee.go:131]: add a description pls, also many functions in this file also doesn't have one. I'm generally okay with that, but don't know how other members will think about it.

File: cli/config/config.go [hunk 3]
```diff
@@ -13,14 +13,16 @@ type Config struct {
 // tt:
 //   modules:
 //     directory: path/to
-// app:
+//   app:
```
Comment (better0fdead) [inline cli/config/config.go:16]: Please move into a different commit

File: cli/search/search.go [hunk 4]
```diff
@@ -37,11 +40,28 @@ func SearchVersions(program string) error {
 	} else if program == "tt" {
 		cmd = exec.Command("git", "-c", "versionsort.suffix=-", "ls-remote", "--tags", "--sort="+
 			"v:refname", "https://github.com/tarantool/tt.git")
+	} else if program == "tarantool-ee" {
```
Comment (better0fdead) [inline cli/search/search.go:43]: I think that it would be better to move Tarantool-ee search into different function and call it here.

File: cli/util/util.go [hunk 5]
```diff
@@ -306,3 +308,50 @@ func AskConfirm(question string) (bool, error) {
...
+			continue
+		}
+
+		res[re.SubexpNames()[i]] = value
+	}
+
+	return res
+}
+
+// Max eturns the maximum value.
```
Comment (better0fdead) [inline cli/util/util.go:350]:
```suggestion
// Max returns the maximum value.
```

File: cli/install_ee/install_ee.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
+package install_ee
```
Comment (better0fdead) [inline cli/install_ee/install_ee.go:1]: I am not sure about that naming, probably need to discuss in a voice call.

File: cli/util/util.go [hunk 7]
```diff
@@ -306,3 +308,50 @@ func AskConfirm(question string) (bool, error) {
...
+	if err != nil {
+		return "", err
+	}
+
+	return strings.TrimSpace(string(out)), nil
+}
+
+// AtoiUint64 converts string to uint64.
+func AtoiUint64(str string) (uint64, error) {
+	res, err := strconv.ParseUint(str, 10, 64)
```
Comment (better0fdead) [inline cli/util/util.go:324]: Pls, merge such condition in one-liners.

File: cli/version/version_tools.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+	TypeRC
+	TypeRelease
+)
+
+type Release struct {
+	Type ReleaseType
+	Num  uint64
+}
+
+type Version struct {
```
Comment (better0fdead) [inline cli/version/version_tools.go:26]: add descriptions to structs, also in my opinion maybe rename to TarVersion or TarantoolVersion or rename module to tar_version_tools to prevent any confusions in future.

**Review by better0fdead — COMMENTED** [review]
Hi, thanks for patch.

**Review by better0fdead — CHANGES_REQUESTED** [review]
I don't see an issue in commit. I think it is follow up for #95.

File: cli/version/version_tools.go (hunk 8 above)
Comment (better0fdead) [reply cli/version/version_tools.go:26]: discussed about name, have no questions

File: cli/config/config.go (hunk 1 above)
Comment (psergee) [inline cli/config/config.go:34]:
```suggestion
// EEOpts is used to store tarantool-ee options.
```

File: cli/config/config.go [hunk 9]
```diff
@@ -29,6 +31,12 @@ type ModulesOpts struct {
 	Directory string
 }
 
+// EEOpts is uses to store tarantool-ee options.
+type EEOpts struct {
+	// Path to file with credentials for downloading tarantool-ee.
```
Comment (psergee) [inline cli/config/config.go:36]:
```suggestion
	// CredPath is a path to file with credentials for downloading tarantool-ee.
```

File: cli/install_ee/install_ee.go [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+	"github.com/tarantool/tt/cli/config"
+	"github.com/tarantool/tt/cli/util"
+	"github.com/tarantool/tt/cli/version"
+	"golang.org/x/term"
+)
+
+const eeSource string = "https://download.tarantool.io/enterprise/"
+
+type creds struct {
```
Comment (psergee) [inline cli/install_ee/install_ee.go:23]: Suppose, this is credentials, not creds. These words have different meaning. loginCredentials or userCredentials will be more meaningful.

File: cli/install_ee/install_ee.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+func getCreds(cliOpts *config.CliOpts) (creds, error) {
+	if cliOpts.EE == nil || (cliOpts.EE != nil && cliOpts.EE.CredPath == "") {
+		return getCredsInteractive()
+	}
+
+	return getCredsFromFile(cliOpts.EE.CredPath)
+}
+
+func getTarballName(data string) (string, error) {
```
Comment (psergee) [inline cli/install_ee/install_ee.go:77]: Please add a comment about what this function is supposed to do.

File: cli/install_ee/install_ee.go [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+	parsedData := re.FindStringSubmatch(data)
+	if len(parsedData) == 0 {
+		return "", fmt.Errorf("Cannot parse tarball name")
+	}
+
+	return parsedData[1], nil
+}
+
+// parseIndex collects a list of versions for the host architecture.
```
Comment (psergee) [inline cli/install_ee/install_ee.go:88]: What versions? Tarantool EE?
What is "data" parameter supposed to contain? Why is this called "parseIndex" and not "getAvailableVersions" or something like that?

File: cli/util/util.go [hunk 13]
```diff
@@ -306,3 +308,50 @@ func AskConfirm(question string) (bool, error) {
...
+		return 0, err
+	}
+
+	return res, nil
+}
+
+// FindNamedMatches processes regexp with named capture groups
+// and transforms output to a map. If capture group is optional
+// and was not found, map value is empty string.
+func FindNamedMatches(re *regexp.Regexp, str string) map[string]string {
```
Comment (psergee) [inline cli/util/util.go:335]: Add unit-test for this API.

File: cli/install_ee/install_ee.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+func parseIndex(data *[]byte) ([]version.Version, error) {
```
Comment (psergee) [inline cli/install_ee/install_ee.go:89]: Add unit-test and description.

File: cli/install_ee/install_ee.go [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+	bytePass, err := term.ReadPassword(int(syscall.Stdin))
+	if err != nil {
+		return res, err
+	}
+	res.password = strings.TrimSpace(string(bytePass))
+
+	return res, nil
+}
+
+func getCredsFromFile(path string) (creds, error) {
```
Comment (psergee) [inline cli/install_ee/install_ee.go:49]: Add unit-test.

File: cli/search/search.go (hunk 4 above)
Comment (0x501D) [reply cli/search/search.go:43]: What for? it called just after
```go
log.Warn("Available versions of " + program + ":")
```
and returns. So we don't need to duplicate code.

File: cli/config/config.go (hunk 1 above)
Comment (0x501D) [reply cli/config/config.go]: Fixed

File: cli/config/config.go (hunk 3 above)
Comment (0x501D) [reply cli/config/config.go:16]: Done

File: cli/install_ee/install_ee.go (hunk 2 above)
Comment (0x501D) [reply cli/install_ee/install_ee.go:137]: Done

File: cli/config/config.go (hunk 1 above)
Comment (0x501D) [reply cli/config/config.go]: Fixed

File: cli/config/config.go (hunk 9 above)
Comment (0x501D) [reply cli/config/config.go]: fixed

File: cli/install_ee/install_ee.go (hunk 12 above)
Comment (0x501D) [reply cli/install_ee/install_ee.go]: Renamed to getVersions.
Comment edited.

File: cli/util/util.go (hunk 5 above)
Comment (0x501D) [reply cli/util/util.go]: Fixed

**Review by better0fdead — APPROVED** [review]

File: cli/util/util.go (hunk 13 above)
Comment (0x501D) [reply cli/util/util.go:335]: Added

File: cli/install_ee/install_ee.go (hunk 15 above)
Comment (0x501D) [reply cli/install_ee/install_ee.go]: Added.

File: cli/install_ee/install_ee.go (hunk 10 above)
Comment (0x501D) [reply cli/install_ee/install_ee.go]: Fixed

File: cli/install_ee/install_ee.go [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+)
+
+const eeSource string = "https://download.tarantool.io/enterprise/"
+
+type userCredentials struct {
+	username string
+	password string
+}
+
+// getCredsInteractive Interactively prompts the user for credentials.
```
Comment (psergee) [inline cli/install_ee/install_ee.go:28]:
```suggestion
// getCredsInteractive interactively prompts the user for credentials.
```

File: cli/install_ee/install_ee.go [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+	if len(parsedData) == 0 {
+		return "", fmt.Errorf("Cannot parse tarball name")
+	}
+
+	return parsedData[1], nil
+}
+
+// getVersions collects a list of all available tarantool-ee
+// versions for the host architecture.
+func getVersions(data *[]byte) ([]version.Version, error) {
```
Comment (psergee) [inline cli/install_ee/install_ee.go:94]: What is the format of input data? Variable name "data" does not give any clue and there is no description in comment. Is it HTML, or JSON?

**Review by psergee — APPROVED** [review]

File: cli/install_ee/install_ee.go (hunk 17 above)
Comment (0x501D) [reply cli/install_ee/install_ee.go:94]: For now it is HTML, but it will be changed in nearest future, so it called `data` for now.
